### PR TITLE
Add support for response_type id_token

### PR DESCRIPTION
**Pull request by januszm**
Current implementation will always trigger the `failure` action in the client for the Implicit grant. It's because we call `Omniauth::Strategy` `.fail!` method if parameter `code` is missing.

Resolves: https://github.com/m0n9oose/omniauth_openid_connect/issues/31

For the Implicit grant, the `id_token` or both `id_token` and `access_token` may be returned.
NOTICE: when `response_type` is set to `token`, then `access_token` parameter is used (Bearer token).

A much simpler implementation to handle just the 2 cases, `code` or `id_token` could look like:

```ruby
def callback_phase
  # ...
  return fail!(:missing_code, OmniAuth::OpenIDConnect::MissingCodeError.new(params['error'])) unless valid_code?
  return fail!(:missing_id_token, OmniAuth::OpenIDConnect::MissingIdTokenError.new(params['error'])) unless valid_id_token?
  # ...
end

def valid_code?
  options.response_type == 'code' ? params.key?('code') : true
end
def valid_id_token?
  options.response_type == 'id_token' ? params.key('id_token') : true
end
```

Let me know which option would you prefer to start with

### CODE REVIEW

**Comment by januszm** [comment]
I didn't have a chance and I won't test the `token`/`access_token`-only scenario unfortunately.

**Comment by januszm** [comment]
This looks like a minor version update, new feature.

**Comment by januszm** [comment]
Hmm we need to return from the `callback_phase` just before:

```ruby
client.authorization_code = authorization_code
access_token
super
```

Maybe with:

```ruby
return something unless options.response_type.to_s == 'code'
```

And for Implicit Grant, `super` will execute:
```ruby
def callback_phase
  env['omniauth.auth'] = auth_hash
  call_app!
end

# =>

def auth_hash
  hash = AuthHash.new(:provider => name, :uid => uid)
  hash.info = info unless skip_info?
  hash.credentials = credentials if credentials
  hash.extra = extra if extra
  hash
end
```


so we need to deal with quite a few things, like

```ruby
def user_info
  @user_info ||= access_token.userinfo!
end
```

File: lib/omniauth/strategies/openid_connect.rb [hunk 1]
```diff
@@ -14,6 +14,8 @@ class OpenIDConnect
       include OmniAuth::Strategy
       extend Forwardable
 
+      SUPPORTED_GRANT_TYPES = %w[code id_token token].freeze
```
Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: why did you add

```ruby
SUPPORTED_GRANT_TYPES = %w[code id_token token]
```

if 

> currently, only `:code` (Authorization Code grant) and `:id_token` (Implicit grant) are valid

?

File: lib/omniauth/strategies/openid_connect.rb [hunk 2]
```diff
@@ -35,7 +37,7 @@ class OpenIDConnect
       option :client_jwk_signing_key
       option :client_x509_signing_key
       option :scope, [:openid]
-      option :response_type, 'code'
+      option :response_type, 'code' # ['code', 'id_token', 'id_token token']
```
Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: same question here. Why do we need `'id_token token'` option?

File: lib/omniauth/strategies/openid_connect.rb [hunk 3]
```diff
@@ -109,7 +111,8 @@ def callback_phase
 
         raise CallbackError, 'Invalid state parameter' if invalid_state
 
-        return fail!(:missing_code, OmniAuth::OpenIDConnect::MissingCodeError.new(params['error'])) unless params['code']
+        grant_type_errors = validate_grant_type_credentials
+        return fail!(*grant_type_errors) if grant_type_errors.present?
```
Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: *Do not* use ActiveSupport's sugar like `#present?`. This gem can be used outside of rails projects.

Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: Sorry, I don't see what's the point to build such complex validation: build an array of errors, then return optional `Array`, splat the array and raise the exceptions. Are you sure this is the easiest way to validate credentials?

I'd prefer to call something like `#validate!` and raise an error inside it, instead of this mess.

Btw, it looks like your validation method supports more than one exception in array. But it will raise just first. Very bad.

File: lib/omniauth/strategies/openid_connect.rb [hunk 4]
```diff
@@ -294,6 +297,18 @@ def logout_path_pattern
         @logout_path_pattern ||= %r{\A#{Regexp.quote(request_path)}(/logout)}
       end
 
+      def validate_grant_type_credentials
+        # It supports the combined 'id_token token' response type
+        options.response_type.to_s.split(' ').each do |grant_type|
+          grant_type_param = grant_type == 'token' ? 'access_token' : grant_type
+          unless params.key?(grant_type_param) && SUPPORTED_GRANT_TYPES.include?(grant_type)
+            error_class = "OmniAuth::OpenIDConnect::Missing#{grant_type.classify}Error".constantize
```
Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: Avoid using Rails methods like `#constantize`

File: lib/omniauth/strategies/openid_connect.rb (hunk 3 above)
Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: Sorry, I was wrong about raising just first exception. It looks like this code will fail if array has more than one exception.

File: lib/omniauth/strategies/openid_connect.rb (hunk 4 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: I actually checked if we use activesupport and it seems that we do, and in such a case it's fine to use `constantize` in my opinion.

`openid_connect` has `active_model` as a runtime dependency, and `active_model` depends on `active_support` (which defines `constantize`).

File: lib/omniauth/strategies/openid_connect.rb (hunk 3 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: It doesn't really matter at this moment , but yeah, I think I'll remove this`validate_grant_type_credentials` , remove support for the combined `id_token token` response type and just focus on making it working with one type, `id_token` in addition to the existing mode

Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: As I mentioned in my other comment, I'm not sure if you are aware but this gem requires active_support anyway and this is the reason why I used both `present?` and `constantize` methods.

File: lib/omniauth/strategies/openid_connect.rb (hunk 2 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: It's one of the valid response types for Implicit Grant.

File: lib/omniauth/strategies/openid_connect.rb (hunk 1 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: I'm going to remove the 'token' response type (together with the combined id_token + token), it's out of scope of my current work and adds too much complexity. I'll update this soon.

File: lib/omniauth/strategies/openid_connect.rb (hunk 4 above)
Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: So you rely on the fact that we have a dependency, which has dependency, which has another dependency.

No, sorry, but this is not fine.

More of that, ActiveSupport does not provide any methods that could be really useful here. `#present` is useful when you don't want to cast object, and can check empty string, `nil`, `false`, etc.. None of that is really need here since you check the result of your own method.

Same with `#constantize`: there's no real reason to dynamically create a constant:

```ruby
CLASSES_MAP = {
  'code' => OmniAuth::OpenIDConnect::MissingCodeError,
  'id_token' => OmniAuth::OpenIDConnect::MissingIdTokenError,
}.freeze

error_class = CLASSES_MAP.fetch(grant_type)
```

Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: Also, the dynamic code is always a bad thing for refactoring. Consider this too.

File: lib/omniauth/strategies/openid_connect.rb (hunk 3 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: This code will actually work, the splat operator simply extracts only two elements of the returned array: message_key and exception. You're right that this code will only return the first exception.
So the question is, if the user (programmer?) specified "id_token foo" as the answer type, should we still try to work with `id_token` (valid) and ignore foo (invalid)? Or should we return an error indicating that this response_type is incorrect as a whole (that was my assumption).

File: lib/omniauth/strategies/openid_connect.rb (hunk 2 above)
Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: I know. But we don't have the full support of it according to the updated readme.

File: lib/omniauth/strategies/openid_connect.rb (hunk 3 above)
Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: Well, since we support only two types `id_token` & `code` I would treat all other values as invalid and show exception as soon as possible.

Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: That's what this code is doing, it's returing one exception as soon as it finds it.

But as I mentioned earlier, I think that we need to take a step back, remove support for the `token` response type and focus on `id_token` only.

Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: I mean there's no need to parse `type` value, try to split it to several values and so far. Just compare it directly, and if there's no match - raise an error.

File: lib/omniauth/strategies/openid_connect.rb (hunk 4 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: The fact is that this gem uses ActiveSupport.
If using these 2 methods are blockers for you, I can remove them.

Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: 1) We can not rely on the fact that we have a dependency indirectly. If you want to use the library you have to add it to dependencies explicitly.
2) There's no real reason to use these methods. I have shown why and how to get rid of them.

These two facts make using these methods inappropriate to me, yes. It would be great if you can get rid of them.

File: lib/omniauth/strategies/openid_connect.rb [hunk 5]
```diff
@@ -207,7 +211,8 @@ def access_token
       end
 
       def decode_id_token(id_token)
-        ::OpenIDConnect::ResponseObject::IdToken.decode(id_token, public_key)
+        # NOTE: we might want to cache the decoded token so it can be used to set basic user info
+        @decoded_id_token ||= ::OpenIDConnect::ResponseObject::IdToken.decode(id_token, public_key)
```
Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: Please, do not memoize result of instance method if the result depends on the passed argument

File: lib/omniauth/strategies/openid_connect.rb [hunk 6]
```diff
@@ -120,6 +121,9 @@ def callback_phase
 
         discover!
         client.redirect_uri = redirect_uri
+
+        return id_token_callback_phase unless options.response_type.to_s == 'code'
```
Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: Please consider changing check to straight logic, instead of reverted

```ruby
return id_token_callback_phase if options.response_type.to_s == 'id_token'
```

Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: I tried both solutions and to be honest I don't have a strong opinion on this one.
The reason why I chose to check for the `code` value is because the rest of the flow depends on the resnpose_type code. If you prefer to stick to `id_token` as the condition for return, I'll change it in next commit.

File: lib/omniauth/strategies/openid_connect.rb (hunk 5 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: Nice catch, I'll use Hash instead. We might want to rename the method as well, since it returns the decodeD token, it can be renamed to `decoded_id_token`. Also, since it takes the parameter, we can pluralize it : `decoded_id_tokens()`

File: lib/omniauth/strategies/openid_connect.rb (hunk 6 above)
Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: This could be a problem if we decide to add more types later.

File: lib/omniauth/strategies/openid_connect.rb (hunk 5 above)
Comment (m0n9oose) [reply lib/omniauth/strategies/openid_connect.rb]: No, please, don't rename this method. It works just fine now and does what it supposed to to do - decode passed token.

File: lib/omniauth/strategies/openid_connect.rb (hunk 6 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: ✅

File: lib/omniauth/strategies/openid_connect.rb (hunk 5 above)
Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: Right, I added another method that uses caching with Hash and calls the original method to do the job

File: lib/omniauth/strategies/openid_connect.rb [hunk 7]
```diff
@@ -210,6 +214,15 @@ def decode_id_token(id_token)
         ::OpenIDConnect::ResponseObject::IdToken.decode(id_token, public_key)
       end
 
+      def decoded_id_tokens(id_token)
```
Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: 1) This method does not return `tokens`! You should not name a method like this just because it also can cache multiple values.

2) Why do we need this caching anyway? Do you have any problems with the performance of this method? This code looks like a great example of premature optimization.

File: lib/omniauth/strategies/openid_connect.rb [hunk 8]
```diff
@@ -109,7 +109,8 @@ def callback_phase
 
         raise CallbackError, 'Invalid state parameter' if invalid_state
 
-        return fail!(:missing_code, OmniAuth::OpenIDConnect::MissingCodeError.new(params['error'])) unless params['code']
+        return fail!(:missing_code, OmniAuth::OpenIDConnect::MissingCodeError.new(params['error'])) unless valid_response_type?('code')
```
Comment (m0n9oose) [inline lib/omniauth/strategies/openid_connect.rb]: Please consider extracting these two lines into separate `#validate` method and rewrite current validation method `#valid_response_type?(type)` to make it more readable:

```ruby
return fail!(:missing_code, OmniAuth::OpenIDConnect::Missing...
return fail!(:missing_code, OmniAuth::OpenIDConnect::Missing...
```

⬇️ 

```ruby
return unless valid_request_type?
```

and

```ruby
def valid_response_type?(type)
  # pass if configured response_type is different than tested here
  options.response_type == type.to_s ? params.key?(type) : true
end
```

⬇️ 

```ruby
TYPES_MAP = {
  'id_token' => { exception_class: OmniAuth::OpenIDConnect::MissingIdTokenError, key: :missing_id_token }.freeze,
  'code' => { exception_class: OmniAuth::OpenIDConnect::MissingCodeError, key: :missing_code }.freeze,
}.freeze

def valid_request_type?
  return true if params.key?(options.response_type)

  # TODO: implement fallback exception
  fallback_exception = nil
  error_attrs = TYPES_MAP.fetch(options.response_type, fallback_exception)
  fail!(error_attrs[:key], error_attrs[:exception_class].new(params['error']))

  false
end
```

Comment (januszm) [reply lib/omniauth/strategies/openid_connect.rb]: Makes sense, it's updated.

**Comment by krzysiek1507** [comment]
It doesn't work when `response_type: %i[id_token code]

```ruby
undefined method `[]' for nil:NilClass
NoMethodError (undefined method `[]' for nil:NilClass):
gems/ruby-2.3.8/bundler/gems/omniauth_openid_connect-3f52ccda27ba/lib/omniauth/strategies/openid_connect.rb:319:in `valid_response_type?'
gems/ruby-2.3.8/bundler/gems/omniauth_openid_connect-3f52ccda27ba/lib/omniauth/strategies/openid_connect.rb:117:in `callback_phase'
omniauth (1.9.0) lib/omniauth/strategy.rb:238:in `callback_call'
```

**Comment by krzysiek1507** [comment]
That's weird that this change was needed. `0.3.2` worked with: `response_type: %i[id_token code]`. I guess id didn't work with `response_type: :id_token`.

**Comment by Eric-Guo** [comment]
it didn't work with `response_type: :code` either

**Comment by m0n9oose** [comment]
> It doesn't work when response_type: %i[id_token code]

Before this change, only `code` was a valid response type. Now it's `code` & `id_token`. So yeah, it's okay if it's not working with multiple values, there's no implementation of that.

> That's weird that this change was needed. 0.3.2 worked with: response_type: %i[id_token code]. I guess id didn't work with response_type: :id_token.

Not really, we were ignoring unknown types.

> it didn't work with response_type: :code either

Will check, thanks!

**Comment by januszm** [comment]
Changes requested in the code review did not include handling both symbol and string version of the exception key for invalid response_type param, here:

`error_attrs = RESPONSE_TYPE_EXCEPTIONS[options.response_type]`
we're not casting the response_type option into string (above, we define exception key as a string (`'id_token' => { exception_class: OmniAuth...`)

**Comment by januszm** [comment]
@krzysiek1507  @Eric-Guo  try this: https://github.com/m0n9oose/omniauth_openid_connect/pull/35 , support for both symbol and string was lost during the code review phase for this PR.

**Comment by krzysiek1507** [comment]
> Before this change, only code was a valid response type. Now it's code & id_token. So yeah, it's okay if it's not working with multiple values, there's no implementation of that.

But it worked well! I need it so let me try to implement it.

**Comment by Eric-Guo** [comment]
Still have below error:

```log

318 319  | error_attrs = RESPONSE_TYPE_EXCEPTIONS[options.response_type]         fail!(error_attrs[:key], error_attrs[:exception_class].new(params['error']))

/Users/guochunzhong/git/oss/omniauth_openid_connect/lib/omniauth/strategies/openid_connect.rb:319:in `valid_response_type?' 
/Users/guochunzhong/git/oss/omniauth_openid_connect/lib/omniauth/strategies/openid_connect.rb:117:in `callback_phase' 

```

**Comment by m0n9oose** [comment]
> Still have below error:
> 
> ```
> 
> 318 319  | error_attrs = RESPONSE_TYPE_EXCEPTIONS[options.response_type]         fail!(error_attrs[:key], error_attrs[:exception_class].new(params['error']))
> 
> /Users/guochunzhong/git/oss/omniauth_openid_connect/lib/omniauth/strategies/openid_connect.rb:319:in `valid_response_type?' 
> /Users/guochunzhong/git/oss/omniauth_openid_connect/lib/omniauth/strategies/openid_connect.rb:117:in `callback_phase' 
> ```

make sure you're on latest `master` version. It looks like you still using same version as before.

**Comment by Eric-Guo** [comment]
Now another error:

```log
    def raise_out!
      raise(env['omniauth.error'] || OmniAuth::Error.new(env['omniauth.error.type']))
    end

omniauth (1.9.0) lib/omniauth/failure_endpoint.rb:25:in `raise_out!' 
omniauth (1.9.0) lib/omniauth/failure_endpoint.rb:20:in `call' 
omniauth (1.9.0) lib/omniauth/failure_endpoint.rb:12:in `call' 
omniauth (1.9.0) lib/omniauth/strategy.rb:491:in `fail!' 
/Users/guochunzhong/git/oss/omniauth_openid_connect/lib/omniauth/strategies/openid_connect.rb:319:in `valid_response_type?' 
/Users/guochunzhong/git/oss/omniauth_openid_connect/lib/omniauth/strategies/openid_connect.rb:117:in `callback_phase' 
```

**Comment by Eric-Guo** [comment]
Acturally, the code in the URL (https://portal.test/auth/openid_connect/callback?code=ZvME77cBxeR9jkg6NsvVH-21O1KhJWAaIkGI5tx5fAg&state=db31263029df1c105792dc8632b90a10)

**Comment by krzysiek1507** [comment]
To fix `response_type: :code/'code'` we need to merge https://github.com/m0n9oose/omniauth_openid_connect/pull/34
